### DISCUSSION TITLE
Trouble implementing livecheck

### DISCUSSION THREAD

**Discussion by mietzen** [Homebrew/discussions#4299] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.4-196-g180c89e
ORIGIN: https://github.com/Homebrew/brew
HEAD: 180c89e17426d4fb1562ac1b2bbe0c4c6e1a3cec
Last commit: 8 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2f7330fe49b9138a13e952ba9aeb2f2051902e3c
Core tap last commit: 5 weeks ago
Core tap branch: master
Core tap JSON: 04 Mar 13:20 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: ["--no-quarantine"]
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: N/A
Xcode: 14.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  szip
==> Downloading https://formulae.brew.sh/api/formula.jws.json


Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  elinks
  hdf5@1.8

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  numpy
  pyyaml
  python@3.11

Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
   M Formula/ext4fuse.rb
   M Formula/sshfs.rb

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /opt/homebrew/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/opt/homebrew/sbin:$PATH"' >> ~/.zshrc
```

### Description of issue

Hi, 
I'm trying to create a (my first) Cask for the Thorium Browser. The Mac releases are shuffled between other "special" releases:
https://github.com/Alex313031/Thorium-Special/releases/

I'm now trying to implement a `livecheck`:

```ruby
  livecheck do
    url "https://api.github.com/repos/Alex313031/Thorium-Special/releases"
    regex(/.*MacOS.*/i)
    strategy :json do |json, regex|
      json[].select { |item| item["name"]&.match?(regex) }
                      .map { |item| item["tag_name"] }
    end
  end
```

But it fails with:

```shell
$ brew audit --new-cask thorium-chromium
curl: (22) The requested URL returned error: 404
audit for thorium-chromium: failed
 - exception while auditing thorium-chromium: key not found: :latest
Error: 1 problem in 1 cask detected
```

I guess `https://api.github.com/repos/Alex313031/Thorium-Special/releases` needs a `POST` request but `brew` performs a `GET` request.

I also tried `page_match`, but the assets are not loaded and can't be scanned -.-

```ruby
  livecheck do
    url "https://github.com/Alex313031/Thorium-Special/releases"
    regex(/href=.*?(M\d+(?:[._-]\d+)+).*ARM64\.dmg/i)
    strategy :page_match do |page, regex|
      page.scan(regex).map { |match| match[0] }
    end
  end
```

Any help is highly appreciated :)

Here is a link to my tap: https://github.com/mietzen/homebrew-thorium-cask

**Comment by SMillerDev** [comment]
How is that livecheck different from the `:github_latest` strategy?

**Reply by mietzen** [reply]
The latest version might not always be a MacOS version. At the moment the latest is [Westmere Release - M110.0.5481.178](https://github.com/Alex313031/Thorium-Special/releases/tag/M110.0.5481.178)

A build only for Westmere CPU's on linux.

I also don't find any documentation on the `:github_latest` strategy.

**Reply by SMillerDev** [reply]
You can use https://docs.brew.sh/Brew-Livecheck#git-tags with a regex then

**Reply by mietzen** [reply]
> You can use https://docs.brew.sh/Brew-Livecheck#git-tags with a regex then

Unfortunately the tags also have the same names, they're just the version number. The newest is still M110, the macOS releases don't get their own tags.

**Reply by SMillerDev** [reply]
Maybe you can ask upstream to make a single release for every version instead of splitting it?

**Reply by samford** [reply]
This is one of those situations where it's unfortunately necessary to iterate through releases to check the assets for the file that's used in the cask. This used to be easy when the assets lists for all the releases were included in a repository's `releases` page HTML but now they're fetched asynchronously on load. There are more details on this in my other comment below but suffice to say that neither the `GithubLatest` nor the `Git` strategy are sufficient in this particular context.

**Comment by mietzen** [comment, marked as answer]
@samford is this expected behaviour for json livecheck?

For the answer See Samfords post below 👇

**Reply by samford** [reply]
The reason why you're receiving a 404 when using the `Json` strategy on an api.github.com URL is because livecheck's `#preprocess_url` method is run on the URL and it gets turned into a `.git` URL (`https://api.github.com/repos/Alex313031.git`) which predictably 404s. You can see this if you check livecheck's debug output for this check:

```
$ brew livecheck --debug thorium-chromium

Cask:             thorium-chromium
Livecheckable?:   Yes

URL:              https://api.github.com/repos/Alex313031/Thorium-Special/releases
URL (processed):  https://api.github.com/repos/Alex313031.git
Strategy:         Json
Regex:            /.*MacOS.*/i
curl: (22) The requested URL returned error: 404
Error: thorium-chromium: Unable to get versions
```

There are two internal issues contributing to this:

1. I forgot to add the recent generic strategies like `Json` to the list of strategies where `#preprocess_url` is skipped. I've opened a brew PR to fix this (https://github.com/Homebrew/brew/pull/14890).

2. `#preprocess_url` should skip `api.github.com` URLs regardless. I could have sworn that we had some logic in `#preprocess_url` to handle this scenario but all I can say is that we don't at the moment. The aforementioned PR will resolve this issue but I'll look into special-casing `api.github.com` URLs in `#preprocess_url` when I have a moment.

-----

Once the brew PR is merged (and you run `brew update` to get the change), you should receive a good response from the GitHub API. However, matching URLs from release assets is somewhat involved.

The `releases` response takes the form of an array of release objects, where you have to iterate over each and check their `assets` array for a `browser_download_url` string with a matching dmg file. I've worked up a `livecheck` block for this, so I'll create a PR in your tap in a few minutes.

-----

For what it's worth, we specifically avoid checking api.github.com in first-party `livecheck` blocks because we've run into rate limiting issues in the past (and livecheck can't handle authentication in this scenario). Instead, we've been using a complicated `strategy` block to iterate through releases, even though it's less efficient (e.g., [`wine-stable`](https://github.com/Homebrew/homebrew-cask/blob/master/Casks/wine-stable.rb#L14-L38)). It's something that I intend to revisit but that's the situation for now.

However, api.github.com is likely fine for your tap, since you only have one cask to worry about (instead of thousands). Anyway, I only mention this in case you run into failures related to rate limiting somewhere down the road.

**Reply by mietzen** [reply]
As already said, thank you for your support! I marked this question as answered.
